src/lib/shmem/src/shmalloc_impl.rs

```diff
@@ -0,0 +1,739 @@
+//! This module implements a low-level, unsafe shared memory allocator that uses mmap'ed shared
+//! memory files as the backing store. The module is intended to be no-std so it can be used in
+//! Shadow's shim library, which must async-signal-safe.
+//!
+//! The allocator chains together chunks of shared memory and divvies out portions of each chunk
+//! using a first-fit strategy. The allocator also implements a freelist so that allocated blocks
+//! can be reused efficiently after free. The allocator design isn't good for general-purpose
+//! allocation, but should be OK when used for just a few types.
+//!
+//! This code is intended to be private; the `allocator` module is the public, safer-to-use
+//! front end.
+
+use crate::raw_syscall::*;
+use numtoa::NumToA;
+
+const PATH_MAX_NBYTES: usize = 255;
+pub(crate) const PATH_BUF_NBYTES: usize = PATH_MAX_NBYTES + 1;
+pub(crate) type PathBuf = [u8; PATH_BUF_NBYTES];
+
+use vasi::VirtualAddressSpaceIndependent;
+
+fn get_null_path_buf() -> PathBuf {
+    [0; PATH_BUF_NBYTES]
+}
+
+#[cfg(debug_assertions)]
+type CanaryBuf = [u8; 4];
```

Comment:
We don't need to do it now, but maybe we could unify this with [`Magic`](https://shadow.github.io/docs/rust/shadow_rs/utility/struct.Magic.html). I think we'd need to move it to a different crate  though, and make it `VirtualAddressSpaceIndependent`, which might mean not storing `TypeId` directly.

Reply:
Leaving this as a todo